ioctl: nvme_set_features use nvme_passthru_cmd directly

MaisenbacherD opened the pull request:
Drop struct nvme_set_features_args.

MaisenbacherD committed:
ioctl: nvme_set_features use nvme_passthru_cmd directly

Drop struct nvme_set_features_args.

Signed-off-by: Dennis Maisenbacher <dennis.maisenbacher@wdc.com>

igaw commented on src/nvme/ioctl.h:
```diff
@@ -2157,12 +2159,67 @@ static inline int nvme_get_log_lockdown(nvme_link_t l,
+				     __u32 timeout, __u32 nsid, __u32 cdw11,
+				     __u32 cdw12, __u32 cdw13, __u32 cdw15,
+				     __u32 data_len, bool save, __u8 uuidx,
+				     __u8 fid)
```
Just a quick review I'll look into to it more tomorrow.

Drop the timeout argumen, in the most cases it is the default anyway. if a user want's to use a different timeout, it should be done via the link object.

Also move the result argument as last, it's the return value.

MaisenbacherD replied:
Reordering of arguments is done. I will still wire up the timeout via the link object. :)

igaw replied:
okay, so you are saying we should do something like

```c
.timeout_ms	= nvme_link_get_timeout(l),
```

?

I was thinking if the submitting command has a timeout value of 0 the passthru function just uses the value from the link. This would avoid a lot of duplicated code on this level.

MaisenbacherD replied:
After some iterations I ended up doing what you were thinking of :)

MaisenbacherD replied:
Let me know if you like the timeout handling now :)

igaw commented on src/nvme/ioctl.h:
```diff
@@ -2157,12 +2159,66 @@ static inline int nvme_get_log_lockdown(nvme_link_t l,
+static inline int nvme_set_features(nvme_link_t l, void *data, __u32 nsid,
+				     __u32 cdw11, __u32 cdw12, __u32 cdw13,
+				     __u32 cdw15, __u32 data_len, bool save,
+				     __u8 uuidx, __u8 fid, __u32 *result)
```
Could you reorder these arguments (see comment in https://github.com/linux-nvme/libnvme/issues/495). And also rename uuidx do `uidx` as it is in the spec? My aim is to follow the spec as closely as possible.

MaisenbacherD replied:
Yes :)

MaisenbacherD replied:
Done :)
I went through all functions I changed and checked against the specs. Please let me know if you spot another mismatch :)

igaw commented on src/nvme/ioctl.h:
```diff
@@ -2387,7 +2502,15 @@ int nvme_set_features_auto_pst(nvme_link_t l, bool apste, bool save,
-int nvme_set_features_timestamp(nvme_link_t l, bool save, __u64 timestamp);
+static inline int nvme_set_features_timestamp(nvme_link_t l, bool save, __u64 timestamp)
+{
+	__le64 t = htole64(timestamp);
```
After checking the spec again and looking a bit closer, this is correct. Because this is the `Data Structure for Set Features` data structure which is not touched by the transport driver, see `nvme_user_cmd`.

MaisenbacherD committed:
ioctl: drop timeout arguments

Introduce a timeout variable to nvme_link that can be adjusted by the
user. If the timeout value of the command to be submitted matches the
default value, it is overwritten with the nvme_link specific timeout.

Signed-off-by: Dennis Maisenbacher <dennis.maisenbacher@wdc.com>

MaisenbacherD commented:
I am working on reordering the arguments now and then providing a nvme-cli patch for those changes.

MaisenbacherD commented:
Reordering of arguments is done. I have some more indentation cleanup to do.

MaisenbacherD commented:
@igaw Do you already have some patches lying around for nvme-cli3 that compiles with current nvme-experiments/libnvme2? Otherwise, I will rebase to where nvme_passthu_cmd was not used directly yet. :)

igaw commented:
nvme-cli3 is in sync up to the ioctl.h changes. so after that it wont compile

igaw commented:
I'll pull this PR and do some reorder of the patches in the branch as rebase etc.

igaw commented:
Thanks btw!

MaisenbacherD commented:
> nvme-cli3 is in sync up to the ioctl.h changes. so after that it wont compile

@igaw nvme-cli3 is currently in sync up to the ioctl.h changes until https://github.com/nvme-experiments/libnvme/commit/d59c02166a96e5d35a4b0fe2e161f15f3153e385

But e.g. https://github.com/nvme-experiments/libnvme/commit/5fa5c8a83d1017e4af65866f906130ccb61e0ebf introduces API changes that are not updated on nvme-cli3 if I am not completely off.
